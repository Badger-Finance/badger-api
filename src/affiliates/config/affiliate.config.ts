import { BadRequest } from '@tsed/exceptions';
import { Chain } from '../../chains/config/chain.config';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';

type Affiliates = { [protocol: string]: Affiliate };

export abstract class Affiliate {
  private static affiliates: Affiliates = {};

  readonly name: string;

  constructor(name: string) {
    this.name = name;
  }

  static register(protocol: Protocol, affiliate: Affiliate): void {
    this.affiliates[protocol] = affiliate;
  }

  static getAffiliate(affiliateData: SettAffiliateData): Affiliate {
    const affiliate = this.affiliates[affiliateData.protocol];
    if (!affiliate) {
      throw new BadRequest(`${affiliate} is not a affiliate`);
    }
    return affiliate;
  }

  abstract getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData>;
}
