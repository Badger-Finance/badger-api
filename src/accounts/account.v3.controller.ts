import { Network } from '@badger-dao/sdk';
import { Controller, Get, Inject, QueryParams } from '@tsed/common';
import { ContentType, Description, Returns, Summary } from '@tsed/schema';
import { Chain } from '../chains/config/chain.config';
import { AccountsService } from './accounts.service';
import { AccountModel } from './interfaces/account-model.interface';
import { QueryParamError } from '../errors/validation/query.param.error';

@Controller('/account')
export class AccountV3Controller {
  @Inject()
  accountsService!: AccountsService;

  @Get()
  @ContentType('json')
  @Summary('Get badger user account information')
  @Description(
    'Return key user information for a given account. Includes positions, earnings from use, and claimable balances.',
  )
  @Returns(200, AccountModel)
  @Returns(400).Description('Not a valid chain')
  @Returns(404).Description('Not a valid account')
  async getAccount(
    @QueryParams('accountId') userId: string,
    @QueryParams('chain') chain: Network,
  ): Promise<AccountModel> {
    if (!userId) throw new QueryParamError('userId');

    return this.accountsService.getAccount(Chain.getChain(chain), userId);
  }
}
