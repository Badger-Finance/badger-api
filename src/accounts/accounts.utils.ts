import { Network } from '@badger-dao/sdk';
import { ethers } from 'ethers';
import { GraphQLClient } from 'graphql-request';
import { getDataMapper } from '../aws/dynamodb.utils';
import { Chain } from '../chains/config/chain.config';
import { TOKENS } from '../config/tokens.config';
import {
  getSdk,
  OrderDirection,
  User_OrderBy,
  UserQuery,
  UserSettBalance,
  UsersQuery,
} from '../graphql/generated/badger';
import { LeaderBoardType } from '../leaderboards/enums/leaderboard-type.enum';
import { CachedBoost } from '../leaderboards/interface/cached-boost.interface';
import { getPrice, inCurrency } from '../prices/prices.utils';
import { getCachedSett, getSettDefinition } from '../setts/setts.utils';
import { cachedTokenBalanceToTokenBalance, formatBalance, getSettTokens, getToken } from '../tokens/tokens.utils';
import { Account } from './interfaces/account.interface';
import { CachedAccount } from './interfaces/cached-account.interface';
import { CachedSettBalance } from './interfaces/cached-sett-balance.interface';

export function defaultBoost(address: string): CachedBoost {
  return {
    leaderboard: LeaderBoardType.BadgerBoost,
    rank: 0,
    address,
    boost: 1,
    stakeRatio: 0,
    nftMultiplier: 0,
    nativeBalance: 0,
    nonNativeBalance: 0,
  };
}

export async function getUserAccount(chain: Chain, accountId: string): Promise<UserQuery> {
  const badgerGraphqlClient = new GraphQLClient(chain.graphUrl);
  const badgerGraphqlSdk = getSdk(badgerGraphqlClient);
  return badgerGraphqlSdk.User({
    id: accountId.toLowerCase(),
    orderDirection: OrderDirection.Asc,
  });
}

export async function getUserAccounts(chain: Chain, accounts: string[]): Promise<UsersQuery> {
  const badgerGraphqlClient = new GraphQLClient(chain.graphUrl);
  const badgerGraphqlSdk = getSdk(badgerGraphqlClient);
  return badgerGraphqlSdk.Users({
    where: {
      id_in: accounts.map((acc) => acc.toLowerCase()),
    },
  });
}

export async function getAccounts(chain: Chain): Promise<string[]> {
  const badgerGraphqlClient = new GraphQLClient(chain.graphUrl);
  const badgerGraphqlSdk = getSdk(badgerGraphqlClient);

  const accounts: string[] = [];

  let lastAddress: string | undefined;
  const pageSize = 1000;
  while (true) {
    try {
      const userPage = await badgerGraphqlSdk.Users({
        first: pageSize,
        where: { id_gt: lastAddress },
        orderBy: User_OrderBy.Id,
        orderDirection: OrderDirection.Asc,
      });
      if (!userPage || !userPage.users || userPage.users.length === 0) {
        break;
      }
      const { users } = userPage;
      lastAddress = users[users.length - 1].id;
      users.forEach((user) => accounts.push(ethers.utils.getAddress(user.id)));
    } catch (err) {
      break;
    }
  }

  return accounts;
}

export function cachedAccountToAccount(cachedAccount: CachedAccount, network?: Network): Account {
  const balances = cachedAccount.balances
    .filter((bal) => !network || bal.network === network)
    .map((bal) => ({
      ...bal,
      tokens: bal.tokens.map((token) => cachedTokenBalanceToTokenBalance(token)),
      earnedTokens: bal.earnedTokens.map((token) => cachedTokenBalanceToTokenBalance(token)),
    }));
  const multipliers = Object.fromEntries(cachedAccount.multipliers.map((entry) => [entry.address, entry.multiplier]));
  const claimableBalances = cachedAccount.claimableBalances.filter((bal) => !network || bal.network === network);
  const claimableBalancesMap = Object.fromEntries(claimableBalances.map((bal) => [bal.address, bal.balance]));
  const account: Account = {
    ...cachedAccount,
    multipliers,
    balances,
    claimableBalances,
    claimableBalancesMap,
  };
  delete account.multipliers[TOKENS.BICVX];
  return account;
}

export async function getCachedAccount(address: string): Promise<CachedAccount> {
  const checksummedAccount = ethers.utils.getAddress(address);
  const defaultAccount = {
    address: checksummedAccount,
    boost: 0,
    boostRank: 0,
    multipliers: [],
    value: 0,
    earnedValue: 0,
    balances: [],
    claimableBalances: [],
    nativeBalance: 0,
    nonNativeBalance: 0,
  };
  try {
    const mapper = getDataMapper();
    for await (const item of mapper.query(
      CachedAccount,
      { address: checksummedAccount },
      { limit: 1, scanIndexForward: false },
    )) {
      return {
        ...defaultAccount,
        ...item,
      };
    }
    return defaultAccount;
  } catch (err) {
    return defaultAccount;
  }
}

export async function toSettBalance(
  chain: Chain,
  settBalance: UserSettBalance,
  currency?: string,
): Promise<CachedSettBalance> {
  const settDefinition = getSettDefinition(chain, settBalance.sett.id);
  const { netShareDeposit, grossDeposit, grossWithdraw } = settBalance;
  const { ppfs } = await getCachedSett(settDefinition);

  const depositToken = getToken(settDefinition.depositToken);
  const settToken = getToken(settDefinition.settToken);
  const currentTokens = formatBalance(netShareDeposit, settToken.decimals);
  let depositTokenDecimals = depositToken.decimals;
  if (depositToken.address === TOKENS.DIGG) {
    depositTokenDecimals = settToken.decimals;
  }
  const depositedTokens = formatBalance(grossDeposit, depositTokenDecimals);
  const withdrawnTokens = formatBalance(grossWithdraw, depositTokenDecimals);
  const balanceTokens = currentTokens * ppfs;
  const earnedBalance = balanceTokens - depositedTokens + withdrawnTokens;
  const [depositTokenPrice, earnedTokens, tokens] = await Promise.all([
    getPrice(settDefinition.depositToken),
    getSettTokens(settDefinition, earnedBalance, currency),
    getSettTokens(settDefinition, balanceTokens, currency),
  ]);

  return Object.assign(new CachedSettBalance(), {
    network: chain.network,
    id: settDefinition.settToken,
    name: settDefinition.name,
    asset: depositToken.symbol,
    ppfs,
    balance: balanceTokens,
    value: inCurrency(depositTokenPrice, currency) * balanceTokens,
    tokens,
    earnedBalance: earnedBalance,
    earnedValue: inCurrency(depositTokenPrice, currency) * earnedBalance,
    earnedTokens,
    depositedBalance: depositedTokens,
    withdrawnBalance: withdrawnTokens,
  });
}

export async function getCachedBoost(address: string): Promise<CachedBoost> {
  const mapper = getDataMapper();
  for await (const entry of mapper.query(
    CachedBoost,
    { address: ethers.utils.getAddress(address) },
    { limit: 1, indexName: 'IndexLeaderBoardRankOnAddress' },
  )) {
    return entry;
  }
  return defaultBoost(address);
}
