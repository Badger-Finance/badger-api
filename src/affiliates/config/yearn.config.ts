import { BigNumber, ethers } from 'ethers';
import { Chain } from '../../chains/config/chain.config';
import { guestListAbi } from '../../config/abi/guest-list.abi';
import { yearnAffiliateVaultWrapperAbi } from '../../config/abi/yearn-affiliate-vault-wrapper.abi';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { getToken } from '../../tokens/tokens.utils';
import { Affiliate } from './affiliate.config';

export class Yearn extends Affiliate {
  constructor() {
    super(Protocol.Yearn);
    Affiliate.register(Protocol.Yearn, this);
  }

  async getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData> {
    const affiliateData: SettAffiliateData = {
      protocol: Protocol.Yearn,
    };
    if (sett.hasBouncer) {
      try {
        const vaultWrapper = new ethers.Contract(sett.settToken, yearnAffiliateVaultWrapperAbi, chain.provider);
        const guestListAddress = await vaultWrapper.guestList();
        const guesList = new ethers.Contract(guestListAddress, guestListAbi, chain.provider);
        const [available, limit]: [BigNumber, BigNumber] = await Promise.all([
          guesList.remainingTotalDepositAllowed(),
          guesList.totalDepositCap(),
        ]);
        const depositToken = getToken(sett.depositToken);
        affiliateData.availableDepositLimit = parseFloat(ethers.utils.formatUnits(available, depositToken.decimals));
        affiliateData.depositLimit = parseFloat(ethers.utils.formatUnits(limit, depositToken.decimals));
      } catch {
        affiliateData.availableDepositLimit = 0;
        affiliateData.depositLimit = 50;
      }
    }
    return affiliateData;
  }
}
