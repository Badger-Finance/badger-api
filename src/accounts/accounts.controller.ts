import { Controller, Get, Inject, PathParams, QueryParams } from '@tsed/common';
import { ContentType, Description, Hidden, Returns, Summary } from '@tsed/schema';
import { Chain } from '../chains/config/chain.config';
import { ChainNetwork } from '../chains/enums/chain-network.enum';
import { AccountsService } from './accounts.service';
import { AccountModel } from './interfaces/account-model.interface';
import { UnclaimedRewards } from './interfaces/unclaimed-rewards.interface';

@Controller('/accounts')
export class AccountsController {
  @Inject()
  accountsService!: AccountsService;

  @Hidden()
  @Get('/allClaimable')
  @ContentType('json')
  async getAllBadgerTreeRewards(@QueryParams('page') page?: number): Promise<UnclaimedRewards> {
    return this.accountsService.getAllUnclaimed(page ?? 1);
  }

  @Get('/:accountId')
  @ContentType('json')
  @Summary('Get badger user account information')
  @Description(
    'Return key user information for a given account. Includes positions, earnings from use, and claimable balances.',
  )
  @Returns(200, AccountModel)
  @(Returns(400).Description('Not a valid chain'))
  @(Returns(404).Description('Not a valid account'))
  async getAccount(
    @PathParams('accountId') userId: string,
    @QueryParams('chain') chain?: ChainNetwork,
  ): Promise<AccountModel> {
    return this.accountsService.getAccount(Chain.getChain(chain), userId);
  }
}
