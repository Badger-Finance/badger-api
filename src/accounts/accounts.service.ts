import { Account } from '@badger-dao/sdk';
import { Service } from '@tsed/di';
import { ethers } from 'ethers';

import { Chain } from '../chains/config/chain.config';
import { InvalidAddrError } from '../errors/validation/invalid.addr.error';
import { getCachedAccount } from './accounts.utils';

@Service()
export class AccountsService {
  async getAccount(chain: Chain, address: string): Promise<Account> {
    let checksumAddress = address;
    try {
      checksumAddress = ethers.utils.getAddress(address);
    } catch {
      throw new InvalidAddrError(`${address}`);
    }
    return getCachedAccount(chain, checksumAddress);
  }
}
