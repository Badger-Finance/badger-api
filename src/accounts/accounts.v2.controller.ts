import { Network } from '@badger-dao/sdk';
import { Controller, Get, Inject, PathParams, QueryParams } from '@tsed/common';
import { ContentType, Deprecated, Description, Returns, Summary } from '@tsed/schema';

import { Chain } from '../chains/config/chain.config';
import { AccountsService } from './accounts.service';
import { AccountModel } from './interfaces/account-model.interface';

@Deprecated()
@Controller('/accounts')
export class AccountsV2Controller {
  @Inject()
  accountsService!: AccountsService;

  @Get('/:accountId')
  @ContentType('json')
  @Summary('Get badger user account information')
  @Description(
    'Return key user information for a given account. Includes positions, earnings from use, and claimable balances.',
  )
  @Returns(200, AccountModel)
  @Returns(400).Description('Not a valid chain')
  @Returns(404).Description('Not a valid account')
  async getAccount(
    @PathParams('accountId') userId: string,
    @QueryParams('chain') chain: Network,
  ): Promise<AccountModel> {
    return this.accountsService.getAccount(Chain.getChain(chain), userId);
  }
}
