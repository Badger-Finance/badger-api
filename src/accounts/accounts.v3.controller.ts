import { Network } from "@badger-dao/sdk";
import { Controller, Get, Inject, QueryParams } from "@tsed/common";
import { ContentType, Description, Returns, Summary } from "@tsed/schema";

import { Chain } from "../chains/config/chain.config";
import { QueryParamError } from "../errors/validation/query.param.error";
import { AccountsService } from "./accounts.service";
import { AccountModel } from "./interfaces/account-model.interface";

@Controller("/accounts")
export class AccountsV3Controller {
  @Inject()
  accountsService!: AccountsService;

  @Get()
  @ContentType("json")
  @Summary("Get badger user account information")
  @Description(
    "Return key user information for a given account. Includes positions, earnings from use, and claimable balances."
  )
  @Returns(200, AccountModel)
  @Returns(400).Description("Not a valid chain")
  @Returns(404).Description("Not a valid account")
  async getAccount(
    @QueryParams("address") address: string,
    @QueryParams("chain") chain?: Network
  ): Promise<AccountModel> {
    if (!address) throw new QueryParamError("address");

    return this.accountsService.getAccount(Chain.getChain(chain), address);
  }
}
