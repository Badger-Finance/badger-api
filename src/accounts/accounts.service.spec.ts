import { Account, Network } from '@badger-dao/sdk';
import { PlatformTest } from '@tsed/common';
import { Ethereum } from '../chains/config/eth.config';
import { LeaderBoardType } from '../leaderboards/enums/leaderboard-type.enum';
import { setupMapper, TEST_ADDR } from '../test/tests.utils';
import { AccountsService } from './accounts.service';
import { CachedAccount } from './interfaces/cached-account.interface';
import * as accountsUtils from './accounts.utils';

describe('accounts.service', () => {
  const chain = new Ethereum();
  let service: AccountsService;
  let result: Account;

  const defaultAccount: CachedAccount = {
    address: TEST_ADDR,
    stakeRatio: 1,
    boost: 2000,
    boostRank: 3,
    multipliers: [],
    nftBalance: 1023,
    value: 320232,
    earnedValue: 2312,
    balances: [
      {
        network: Network.Ethereum,
        address: TEST_ADDR,
        name: 'Example Sett Balance',
        symbol: 'bESB',
        balance: 1,
        value: 320232,
        earnedValue: 2312,
        tokens: [],
        earnedTokens: [],
        pricePerFullShare: 1.002,
        earnedBalance: 1,
        withdrawnBalance: 0,
        depositedBalance: 5,
      },
      {
        network: Network.Arbitrum,
        address: TEST_ADDR,
        name: 'Example Sett Balance',
        symbol: 'bESB',
        balance: 1,
        value: 320232,
        earnedValue: 2312,
        tokens: [],
        earnedTokens: [],
        pricePerFullShare: 1.002,
        earnedBalance: 1,
        withdrawnBalance: 0,
        depositedBalance: 5,
      },
    ],
    claimableBalances: [],
    nativeBalance: 2033222,
    nonNativeBalance: 23129,
  };

  beforeAll(async () => {
    await PlatformTest.create();
    service = PlatformTest.get<AccountsService>(AccountsService);
  });

  beforeEach(async () => {
    setupMapper([defaultAccount]);
    jest.spyOn(accountsUtils, 'getCachedBoost').mockImplementation(async () => ({
      leaderboard: `${chain.network}_${LeaderBoardType.BadgerBoost}`,
      address: TEST_ADDR,
      rank: 3,
      boost: 2000,
      stakeRatio: 1,
      nftBalance: 123213,
      nativeBalance: 2033222,
      nonNativeBalance: 23129,
    }));
    result = await service.getAccount(chain, TEST_ADDR);
  });

  afterEach(PlatformTest.reset);

  describe('getAccount', () => {
    it('returns the expected account', () => {
      expect(result).toMatchSnapshot();
    });
  });
});
