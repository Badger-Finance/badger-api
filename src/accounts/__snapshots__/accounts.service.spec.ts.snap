// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accounts.service getAccount returns the expected account 1`] = `
Object {
  "address": "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484",
  "boost": 2000,
  "boostRank": 1,
  "bveCvxBalance": 1,
  "claimableBalances": Object {},
  "data": Object {},
  "earnedValue": 1,
  "multipliers": Object {},
  "nativeBalance": 5,
  "nftBalance": 3,
  "nonNativeBalance": 5,
  "rank": 1,
  "stakeRatio": 1,
  "value": 10,
}
`;
